pkg/controllers/trace_controller.go

```diff
@@ -0,0 +1,247 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager"
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+)
+
+const (
+	/* Inspired from Gardener
+	 * https://gardener.cloud/documentation/guides/administer_shoots/trigger-shoot-operations/
+	 */
+
+	GADGET_OPERATION = "gadget.kinvolk.io/operation"
+	GADGET_FINALIZER = "gadget.kinvolk.io/finalizer"
+)
+
+// TraceReconciler reconciles a Trace object
+type TraceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Node   string
+
+	GadgetRegistry map[string]gadgets.Gadget
+	TracerManager  *gadgettracermanager.GadgetTracerManager
+}
+
+//+kubebuilder:rbac:groups=gadget.kinvolk.io,resources=traces,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=gadget.kinvolk.io,resources=traces/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=gadget.kinvolk.io,resources=traces/finalizers,verbs=update
+
+func genSelector(f *gadgetv1alpha1.ContainerFilter) *pb.ContainerSelector {
+	if f == nil {
+		return &pb.ContainerSelector{}
+	}
+	labels := []*pb.Label{}
+	for k, v := range f.Labels {
+		labels = append(labels, &pb.Label{Key: k, Value: v})
+	}
+	return &pb.ContainerSelector{
+		Namespace:     f.Namespace,
+		Podname:       f.Podname,
+		Labels:        labels,
+		ContainerName: f.ContainerName,
+	}
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Trace object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *TraceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	trace := &gadgetv1alpha1.Trace{}
+	err := r.Get(ctx, req.NamespacedName, trace)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Infof("Trace %q has been deleted", req.NamespacedName.String())
```

Comment:
```suggestion
			log.Debugf("Trace %q has been deleted", req.NamespacedName.String())
```

Reply:
This would require the Gadget Tracer Manager log levels to be configurable; at the moment they are not, so `log.Debugf` is not printed.

Example of code doing that: https://github.com/kinvolk/traceloop/blob/6f4efc6fca46d92c75f4ec4e6c6e1d829bdeaddf/traceloop.go#L57-L58

Can this be done in a different PR?

Reply:
I've done the changes to `log.Errorf` but not the changes to `log.Debug`.

Reply:
> Can this be done in a different PR?

For sure.